theano/gof/destroyhandler.py

```diff
@@ -956,9 +968,12 @@ def validate(self, fgraph):
                     # graph might have already changed when we raise the
                     # self.fail_validate error. So before raising the error, we
                     # double check here.
-                    for app in app_err_pairs:
-                        if app in fgraph.apply_nodes:
-                            self.fast_destroy(app, 'validate')
+                    #for app in app_err_pairs:
+                    #    if app in fgraph.apply_nodes:
+                    #        self.fast_destroy(app, 'validate')
+                    for app in fgraph.apply_nodes:
+                        self.fast_destroy(app, 'validate')
+                    self.fail_validate = app_err_pairs
```

Comment:
I think we need to document how it work. Here is some self notes that I would like to end up in that doc. For later when we are close to merge as this can still change.

- We needed to take care of cases where we record a possible failure, but it is another feature that cause the revert. 
- We need to only verify nodes that are still in the graph. Nodes marked as failure are potential failure and need to be rechecked during validate()
- We need to handle the case where after a failed validate, there is no revert and the following validate still need to fail.
- In the case of revert, we need to track that a nodes isn't in the graph anymore and we need to remove from potential failures nodes whose inputs aren't in potential failure anymore.

If you think of other cases we try to cover, add reply with them.